missilewars-plugin/src/main/java/de/butzlabben/missilewars/listener/PlayerListener.java

```diff
@@ -92,117 +84,125 @@ public void onPickup(PlayerPickupItemEvent event) {
-                p.teleport(Config.getFallbackSpawn());
-            }
-        }
+        // old game handling:
+        registerPlayerArenaLeaveEvent(player, game);
+
+        game.teleportToFallbackSpawn(player);
     }
 
     @EventHandler
     public void onTeleport(PlayerTeleportEvent event) {
-        Game to = getGame(event.getTo());
-        Game from = getGame(event.getFrom());
-        if (checkJoinOrLeave(event.getPlayer(), from, to)) {
-            event.setCancelled(true);
-        }
+        Location from = event.getFrom();
+        Location to = event.getTo();
+        Game gameFrom = getGame(from);
+        Game gameTo = getGame(to);
+
+        // same game:
+        if (gameFrom == gameTo) return;
+
+        Player player = event.getPlayer();
+
+        // old game handling:
+        if (gameFrom != null) registerPlayerArenaLeaveEvent(player, gameFrom);
+
+        // teleport after a delay between the arena leave and the next area join
+        new BukkitRunnable() {
+            public void run() {
+                // new game handling:
+                if (gameTo != null) {
+                    PlayerArenaJoinEvent joinEvent = registerPlayerArenaJoinEvent(player, gameTo);
+                    if (joinEvent.isCancelled()) gameTo.teleportToFallbackSpawn(player);
+                }
+            }
+        }.runTaskLater(MissileWars.getInstance(), 2);
     }
 
     @EventHandler
     public void onMove(PlayerMoveEvent event) {
-        Game to = getGame(event.getTo());
-        Game from = getGame(event.getFrom());
-
-        if (checkJoinOrLeave(event.getPlayer(), from, to)) {
-            event.setCancelled(true);
-            Vector addTo = event.getFrom().toVector().subtract(event.getTo().toVector()).multiply(3);
-            addTo.setY(0);
-            event.getPlayer().teleport(event.getFrom().add(addTo));
-        }
-
-    }
-
-    @EventHandler(priority = EventPriority.HIGHEST, ignoreCancelled = true)
-    public void onLeaveWithStuff(PlayerArenaLeaveEvent event) {
-        Game game = event.getGame();
+        Location from = event.getFrom();
+        Location to = event.getTo();
+        Game gameFrom = getGame(from);
+        Game gameTo = getGame(to);
 
-        MWPlayer mwPlayer = game.getPlayer(event.getPlayer());
-        Player player = mwPlayer.getPlayer();
+        // same game:
+        if (gameFrom == gameTo) return;
 
-        PlayerDataProvider.getInstance().loadInventory(player);
+        Player player = event.getPlayer();
 
-        game.removePlayer(mwPlayer);
+        // old game handling:
+        if (gameFrom != null) registerPlayerArenaLeaveEvent(player, gameFrom);
 
-        MissileWars.getInstance().getSignRepository().getSigns(game).forEach(MWSign::update);
+        // new game handling:
+        if (gameTo != null) {
+            PlayerArenaJoinEvent joinEvent = registerPlayerArenaJoinEvent(player, gameTo);
+            if (!(joinEvent.isCancelled())) return;
```

Comment:
Remove unnecessary brackets
```suggestion
            if (!joinEvent.isCancelled()) return;
```